### PR TITLE
Adding completion data reporting for moodle

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/edx-enterprise), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

V0 Draft:
What remains
- testing
- adding real doc strings
- decide on possible configurations?

### CODE REVIEW

File: integrated_channels/moodle/client.py
Comment (taliaedX) [inline integrated_channels/moodle/client.py:161]: these jokers really spell it `enrol` and `enrolment`, but then use `enrolled`. _flips table_

File: integrated_channels/moodle/client.py
Comment (taliaedX) [inline integrated_channels/moodle/client.py:279]: what actually is the value of Not found here? I'm just not familiar with it.

File: integrated_channels/moodle/client.py
Comment (taliaedX) [inline integrated_channels/moodle/client.py:344]: I assume you mean the grade on edx's side is a decimal (TIL if so) and we can add logic to pass this to Moodle however the customer would like? Can you even do a letter (ie is the field on Moodle's side expecting an int? Or is it just a char field?

File: integrated_channels/moodle/exporters/learner_data.py
Comment (taliaedX) [inline integrated_channels/moodle/exporters/learner_data.py:36]: I didn't even know about this util. Nice.

**Review by taliaedX — COMMENTED** [review]
I have some questions; mostly nitpicks, but generally looks good.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:279]: 404!

Comment (taliaedX) [reply integrated_channels/moodle/client.py:279]: oh it's just the status code. They don't use `.code`? Weird

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:161]: ...As I'm sure you know already, that's not even the worst part! They have function names misspelled in their docs that don't have a list of parameters so when you grep through the source code nothing comes up 😞

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:224]: this is actually the incorrect way to use `isinstance` 🤦  will change

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:161]: oh yes like grades vs grade

File: integrated_channels/moodle/exporters/learner_data.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/exporters/learner_data.py:36]: (tbh I learned about it while implementing Canvas and seeing some other channel do it)

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:224]: ah yes I missed that. Good follow up/catch!

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:180]: I don't believe that
```
if not user_id:
``` 
will appropriately handle the condition that we never find the user's email in the list of enrolled students without the assignment to None. However, please correct me if I'm wrong! Happy to remove it if it's unnecessary.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:231]: @georgebabey mentioned to me that we might have a contact within Moodle. I think being able to talk to someone who has some insider knowledge will be helpful in choosing the right course of action.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:279]: I just tried `.code` just to __make sure__ aannnnd got an attribute not found.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:344]: I added the percent grade to the exportation, and I believe (some investigation pending?) that things like letter grades should be accessible too. And yeah, the main point of the comment was to say that we could allow the customer to configure this and I'm wondering if we want to add the functionality to this PR

Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:344]: To add more clarity- how I see this configuration (+ creating shell final grade assignment configurations?), we'd probably add another admin form/model for MoodleCompletionDataConfiguration or something similarly named.

File: integrated_channels/moodle/client.py
Comment (binodpant) [inline integrated_channels/moodle/client.py:157]: is this text intentional or actually means the field is a placeholder?

File: integrated_channels/moodle/client.py
Comment (binodpant) [inline integrated_channels/moodle/client.py:176]: ??

File: integrated_channels/moodle/client.py
Comment (binodpant) [inline integrated_channels/moodle/client.py:235]: is this a constant somewhere, or should be if used outside of this code as well?

Comment (binodpant) [reply integrated_channels/moodle/client.py:235]: ah yeah I would constantify it esp since it's used again later here too

File: integrated_channels/moodle/client.py
Comment (binodpant) [reply integrated_channels/moodle/client.py:279]: from their doc:
```>>> HTTPStatus.OK.value
200
>>> HTTPStatus.OK.phrase
'OK'
```

File: integrated_channels/moodle/client.py
Comment (binodpant) [inline integrated_channels/moodle/client.py:341]: I think the `grades[0][FIELD]` needs a brief comment here? it may not be obvious reading this code why grades[0] .. for example or maybe in function docstring is good too. 

also, will there ever be a grades[1] ?

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:180]: I've read that first sentence several times and I don't quite understand.
You have: 
```
if enrollment.get('email') == user_email:
    user_id = enrollment.get('id')
    break
```
so if we never get a user_id from this loop because there is never a match then the next logic:
`if not user_id` would still catch it because user_id is still None. It's never defined. This is what I mean by not needing to set it explicitly to None.

File: integrated_channels/moodle/exporters/learner_data.py
Comment (binodpant) [inline integrated_channels/moodle/exporters/learner_data.py:39]: small but may look cleaner? I feel like since there are so many uses of `enterprise_enrollment.enterprise_customer_user`, a cached variable would be nice ...

File: integrated_channels/moodle/models.py
Comment (binodpant) [inline integrated_channels/moodle/models.py:136]: hmm, grade max_length 100 chars? too long?

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:344]: Yeah but my question was does Moodle support a letter grade?
I ask because their date field (which is an int [commence groaning]) is strictly typed and I wasn't sure if this grade field is also strictly typed since you are currently passing an int. 

TL;DR if you set grade to "B" here would it fail to load into Moodle?

File: integrated_channels/moodle/models.py
Comment (binodpant) [inline integrated_channels/moodle/models.py:131]: would EmailField be better here?

https://docs.djangoproject.com/en/3.1/ref/models/fields/#emailfield

File: integrated_channels/moodle/models.py
Comment (binodpant) [inline integrated_channels/moodle/models.py:139]: minor question I guess, is default True sensible because we are already transmitting grades? is default False any better/worse?

File: integrated_channels/moodle/transmitters/learner_data.py
Comment (binodpant) [inline integrated_channels/moodle/transmitters/learner_data.py:27]: Canvas? :)

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:341]: oh this reminded me actually...there's a spot where I too had array indices to deal with and I did iterate through them in the exporter. That may be what we want here if we do have the potential for multiple grade items in one post.

However, I don't think the format needs a comment, because this kind of formatting nonsense is sprinkled throughout every Moodle client/exporter/transmitter implementation and amounts to "nonsensical and inconsistent PHP array stuff"

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:157]: it means I was rushing to get the PR up and said to myself "oh I'll just add the doc strings soon" 😬

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:165]: yeah, I'm actually confused as to why this is succeeding... it should be a post. Agreed on the usage of the generalized wrapper.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:180]: Resolved offline! I believe this comes down to preferences of implementation. There are ways around the predefining of `user_id` to None, but I think my solution is perfectly reasonable here.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:165]: Tested locally- Seems like making both a post and get request to Moodle succeed and return the expected list of user enrollments? Coooooool. In that case, since we already have the `_post` wrapper method, perhaps I should just use that?

Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:165]: yeah, this entire method can be simplified down to 
```
    @moodle_request_wrapper
    def _get_enrolled_users(self, course_id):
        """
        placeholder
        """
        params = {
            'wstoken': self.token,
            'wsfunction': 'core_enrol_get_enrolled_users',
            'courseid': course_id,
            'moodlewsrestformat': 'json'
        }
        return self._post(params)
```

Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:165]: (finally taking the placeholder out now)

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:180]: This was also my misunderstanding. I have used `if not var` in a lot of places where the variable was not defined in the given code block, but didn't think about the fact that function parameters still count as definition.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:231]: Removed the comment but made sure to bring up with the team the usefulness of talking with Moodle/customers.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:344]: Looking at the UI on Moodle's side, the activities (what the user has to create in order for us to send them grades) don't seem to support letter grades, though there is a _different_ grade "item" (the one I originally thought would be our solution to our implementation) in Moodle that supports a grade type of `text`. It feels like this is how you'd get a letter grade. However having investigated and now discussed with the team, I don't think that configurable options are the best play for this PR. We can communicate with customers going forward and if there's interest then I think we can refer back to this thread.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:157]: annnnd replaced!

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:176]: Fear not! The placeholders are gone!

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:235]: It is not a constant, that's a great point and one that I meant to do but forgot. Good catch, thanks!!

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:341]: Uhg this is really because of Moodle [`tl:dr`](https://github.com/moodle/moodle/blob/99777d963f740ced7920137345a43fdec187da11/lib/classes/grades_external.php#L432) 
The parameters are expected to be an array of grades to update: ie
```
[
   { 
      'studentid': <student ID 1>,
      'grade': <grade to be assigned>
   }
   { 
      'studentid': <student ID 2>,
      'grade': <grade to be assigned>
   }
]
```
So by saying `grades[0][FIELD]` we're creating a single entry for one grade item to be updated. Talia is pretty much spot on, there's the possibility of us doing bulk completion reporting. However, since the client only deals in single user ID's/payloads of only one enrollment, I don't think we can/need to create the ability for the bulk update (without some sort of refactoring of how the client works). I'm happy to explain this as a tl:dr inline comment.

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:161]: 😠  agreed

File: integrated_channels/moodle/models.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/models.py:136]: 🤦 haha yeah, that's probably supposed to be an int/decimal, not even a char

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:341]: Ah yeah in the case I was formatting during the exporter phase, so yeah since we're not formatting this parameter until the client it should be fine. I'm not a huge fan of formatting it in that way (ie I'd prefer we do format in the exporter and can just chunk and send without too much extra formatting), but at that point we would have to use the method I had in the content_metadata exporter.

Comment (taliaedX) [reply integrated_channels/moodle/client.py:341]: also is there a fundamental reason we can't send a bunch of grades at once? I know we've queried around one at a time, but Moodle does support multiple objects/grades unlike Canvas, so we don't necessarily have to chunk at 1.

Comment (taliaedX) [reply integrated_channels/moodle/client.py:341]: ie if you leave the chunk size as its default (which is 100 or 500 or something) and have multiple students and grade items, does this function as is?

File: integrated_channels/moodle/models.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/models.py:139]: Yeah, I have the default here because we have other checks in the transmission flow that make sure the course is completed. I don't think it's necessary either way, but willing to make a change if you think it's better practice to "assume it's not until we prove it is"

File: integrated_channels/moodle/transmitters/learner_data.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/transmitters/learner_data.py:27]: 😠  why is there always one place that I forgot

File: integrated_channels/moodle/client.py
Comment (binodpant) [reply integrated_channels/moodle/client.py:344]: yeah I think we may want to doc this somewhere on what to expect on grading .. or sth like it

File: integrated_channels/moodle/models.py
Comment (binodpant) [reply integrated_channels/moodle/models.py:139]: based on how your code is setup this change is not needed, but yeah as a general dataa modeling concern perhaps it makes sense to default to False? then someone can explicitly set to True if conditions are right. I think a False for this value will mean no grade transmission, so wil leave it up to you

File: integrated_channels/moodle/client.py
Comment (alex-sheehan-edx) [reply integrated_channels/moodle/client.py:335]: Looking at the base transmitter, I think you're right.  it should be able to handle chunk sizes larger than 1, but we would need to make some modifications to the Moodle exporter and transmitter. 

  From the base transmitter-
```
                code, body = self.client.create_course_completion(
                    getattr(learner_data, kwargs.get('remote_user_id')),
                    serialized_payload
                )
```
We would have to make sure that `learner_data['remote_user_id']` is exported as an array of user emails rather than a single one. As you say, and did in with the metadata exporter, the payload formatting (ie `'grades[0][studentid]'`) could be done in the moodle exporter rather than client if we're inclined to go in that direction. Do we want this implemented in this iteration?

**Review by binodpant — APPROVED** [review]
lgtm! if we find any issues we can address them later, but this seems like it will handle one student one grade item just fine

File: integrated_channels/moodle/client.py
Comment (taliaedX) [reply integrated_channels/moodle/client.py:335]: hmmm I think we'd want to support multiple grade items immediately. I wouldn't mind leaving this until later if we had other work we definitely need to do, but since we are _ideally_ at just bug fixes and small features from customer/Ken feedback, I think it'd be best if we support this now

Comment (binodpant) [reply integrated_channels/moodle/client.py:335]: I like the idea of supporting this this, but also end of sprint so, probably a new PR?

Comment (taliaedX) [reply integrated_channels/moodle/client.py:335]: My argument is that I don't know if this even currently works when multiple items are present and a chunk size is not 1.

If we can process them one at a time successfully, then sure we can make a separate pr to support multiple items at once.

I also feel like just because Canvas cannot support multiple things at once, doesn't mean that should necessarily be our standard to call something done but that's just me.

Comment (binodpant) [reply integrated_channels/moodle/client.py:335]: Hmm ok, sorry I probably misunderstood then. If this does not work when chunk size is set to more than 1 (despite moodle supporting multiple items) then yes it's worth fixing. Canvas was different since its api does not support multiple items at once. if that's not true of moodle, agree a fix is worth doing now rather than later

Comment (taliaedX) [reply integrated_channels/moodle/client.py:335]: I meant to update this and forgot - Alex verified it still sends 1 learner's data at a time even with a >1 chunk size.

**Review by taliaedX — APPROVED** [review]
after discussing this all LGTM!

**Pull request merged**
